### PR TITLE
[precompile] Stop the dynamo tracer accumulating into the example's gradients

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #195284
* #195215
* #195070
* #195014
* #194966
* #194952
* #194951
* #194950
* #194949
* #194948
* #194677
* #194622
* #194619
* #194549
* #194542
* #194541
* #194534
* #194488
* #194479
* #194455
* #194454
* #194452
* #194450
* #194449
* #194428
* #194427
* __->__ #194426
* #194409
* #194394
* #194393
* #194390
* #194389
* #194388
* #194387
* #194386
* #194385
* #194384
* #194383
* #194320
* #194319
* #194318
* #194317
* #194316
* #194302
* #194301
* #194300
* #194286
* #194285
* #194284
* #194283
* #194282
* #194250
* #194249
* #194248
* #194158
* #192379
* #193976
* #193697
* #192374
* #192866

``torch._precompile`` states of its capture path that "precompile does not
mutate the user's example .grad (params/buffers AND user inputs)", and the
make_fx tracer implements it: snapshot the ORIGINAL grad object, clear, restore
the same object in a finally, so both value and identity survive. The dynamo
tracer runs its examples directly and never did any of that.

A training example runs a real backward, which ACCUMULATES. So capturing a model
that had already taken a step doubled its gradients:

    one eager step   |w.grad| = 14.351556
    after precompile |w.grad| = 28.703112     ratio 2.0

That is the documented warmup-step-then-capture flow, and the damage does not
stop at the capture: the doubled gradient goes through the optimizer, so every
subsequent step of the run is computed from perturbed parameters. On a
capture-then-train loop with SGD+momentum at lr=1e-4 the batch after the capture
comes out 1.7e-4 relative from an uncaptured baseline -- small enough to read as
inductor reassociation, which is exactly how it was reported to us.

Snapshot before the examples run and restore in the finally, the same shape
make_fx uses, over every tensor an example could accumulate into: the tensor
arguments, the parameters and buffers of any nn.Module argument, and the bound
receiver when the captured callable is a method. The same grad OBJECT goes back,
not a copy, because optimizer state can be keyed on its identity.

Only the automatic ``example_inputs`` are covered. Calls the caller makes inside
the ``with`` block of the session API are the caller's own and run in the
ambient grad mode by documented design.

Test Plan:

```
python test/test_precompile.py
python test/dynamo/test_package.py
python test/dynamo/test_guard_serialization.py
```

The test that covered this is rewritten. It asserted that a capture accumulates
ONCE rather than twice, which blessed the mutation instead of forbidding it; it
now asserts the gradients are unchanged and the grad object is identical, and it
is parametrized over both tracers -- the three existing tests that pin this
contract all call precompile() without ``tracer=``, so they only ever exercised
the path that already honoured it.

Authored with an AI assistant.

Differential Revision: [D117030053](https://our.internmc.facebook.com/intern/diff/D117030053)

### CODE REVIEW

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D117030053).

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D117030053).

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D117030053).

**Comment by bobrenjc93** [comment]
Superseded by the canonical precompile stack #195353-#195362; this change lives in #195362 (with the guard-drop policy half in #195360). Closing to consolidate review there. The branch is kept on GitHub.
